Add breaking change policy to website (#6049)
microsoft/typespec: website/src/content/docs/docs/handbook/breaking-change-policy.mdx

Author: add the stdout of the compiler is not a contract, use teh api for programmatic usage

---
title: Breaking Change Policy
---

# Breaking Change Policy

import { Badge } from "@astrojs/starlight/components";

:::note
This is to be applicable after 1.0.0 release
:::

## Versioning Scheme

> TypeSpec Compiler and libraries follow [semver](https://semver.org/) for versioning. Which means a breaking change(according the following policy) will only be introduced in a major version release.

Reviewer: Honestly I'd leave this out. Semver is, I believe, quite prescriptive about what constittutes a breaking change, and stating this invites lawyering about that definition. Consider maybe: we follow a major, minor, patch versioning scheme according to the following policy:

## Breaking Change Philosophy

> The basic of the policy can be summarized as follows, we guarantee stability for runtime behavior but not for the build process.

Reviewer: ```suggestion
The basic of the policy can be summarized as follows: existing language syntax and semantics will not change without a major version. The runtime behavior for a given TypeSpec will not change without a major version. TypeSpecs leveraging new or updated features may cause runtime API consumers to fail (e.g. because a new type is unhandled) or builds to fail (e.g. because exhaustive unions are no longer exhaustive).
``

> - A spec that was building successfully should build successfully with the new version of the compiler and libraries with the same major version.

Reviewer: ```suggestion
- The language syntax for existing language elements will not change within a major version. New language elements (which in turn may not be handled by runtime API consumers like emitters) may be added in a minor version.
- A spec that was building successfully should build successfully with the new version of the compiler and libraries with the same major version.
```

- New types and functionalities can be added in a minor version release.
  - If a spec updates to use that new functionality/type it is not considered a breaking change if it goes through a library/emitter that will now break.

> - TypeScript types might change in ways that introcue compilation type checking errors. Those are not considered breaking changes.

Reviewer: ```suggestion
- TypeScript types might change in ways that introduce compilation type checking errors. Those are not considered breaking changes.
```

- <Badge text="Level 1" variant="note" />: Those describe change that are not semantically relevant
  but still produce noticeable changes in the output. This could be for example a change in the way
  the output is formatted.
- <Badge text="Level 2" variant="tip" />: Those are internal apis that could be accessed but
  touching it would be at the library author risk and breaking change wouldn't even need to be
  documented

| Example                          | Level                               | Description                                                                                                                                                                                                                                                                                       |
| -------------------------------- | ----------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | TypeKit                          | <Badge text="0" variant="danger" /> | TypeKit will be moving forward the recommend way to expose apis for a library. Those should be extremely stable.                                                                                                                                                                                  |

Reviewer: ```suggestion
| TypeKit                          | <Badge text="0" variant="danger" /> | TypeKits are the recommend way to expose apis for a library. Those should be extremely stable.                                                                                                                                                                                  |
```

| Additional helpers for a library | <Badge text="0" variant="danger" /> | Depending on the helper this could be considered critical or more advanced usage                                                                                                                                                                                                                  |

> | TypeSpec SyntaxTree              | <Badge text="2" variant="tip" />    | The TypeSpec Ast is not part of the contract a library should depend on.                                                                                                                                                                                                                          |

Reviewer: ```suggestion
| TypeSpec Syntax Tree              | <Badge text="2" variant="tip" />    | The TypeSpec AST is internal and may change at any time.                                                                                                                                                                                                                          |
```

| Checker                          | <Badge text="2" variant="tip" />    | Checker is an internal of the typespec program                                                                                                                                                                                                                                                    |
| Symbols                          | <Badge text="2" variant="tip" />    | Internals                                                                                                                                                                                                                                                                                         |

> | Formatter                        | <Badge text="1" variant="note" />   | Formatting shouldn't' affect the meaning of the language. Changes to the formatting should be taken with consideration as it could create a lot of diff but it shouldn't also be considered a breaking change                                                                                     |

Reviewer: ```suggestion
| Formatter                        | <Badge text="1" variant="note" />   | Formatting shouldn't affect the meaning of the language. Changes to the formatting should be taken with consideration as it could create a lot of diff but it shouldn't also be considered a breaking change                                                                                     |
```

| Bug                              | [See bug](#bugs)                    | I think if something is found to be a bug from the documented behavior and fixing it is in someway breaking it shouldn't be considered as such unless it is such a widely used item where it has to become a "feature". Those should however still be documented as breaking in the release notes |

### Supported node version

> TypeSpec support all currently supported node version by the node foundation.

Reviewer: ```suggestion
TypeSpec intends to support all [current node.js releases](https://nodejs.org/en/about/previous-releases). We may drop support for a maintenance or unsupported node version within a major version.
```

---
title: Breaking Change Policy
---

# Breaking Change Policy

import { Badge } from "@astrojs/starlight/components";

:::note
This is to be applicable after 1.0.0 release
:::

## Versioning Scheme

> TypeSpec Compiler and libraries follow [semver](https://semver.org/) for versioning. Which means a breaking change(according the following policy) will only be introduced in a major version release.

Author: ```suggestion
TypeSpec Compiler and libraries follow a major, minor, patch versioning scheme according to the following policy:
```

- <Badge text="Level 1" variant="note" />: Those describe change that are not semantically relevant
  but still produce noticeable changes in the output. This could be for example a change in the way
  the output is formatted.
- <Badge text="Level 2" variant="tip" />: Those are internal apis that could be accessed but
  touching it would be at the library author risk and breaking change wouldn't even need to be
  documented

| Example                          | Level                               | Description                                                                                                                                                                                                                                                                                       |
| -------------------------------- | ----------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | TypeKit                          | <Badge text="0" variant="danger" /> | TypeKits are the recommend way to expose apis for a library. Those should be extremely stable.                                                                                                                                                                                  |

Reviewer 2: ```suggestion
| TypeKits                         | <Badge text="0" variant="danger" /> | TypeKits are the recommend way to expose APIs from a library. TypeKits that are only available through experimental subpaths may be broken without a major version update, but TypeKits that are available by default may be considered highly stable.                                                                                                                                                                            |
```

> | Additional helpers for a library | <Badge text="0" variant="danger" /> | Depending on the helper this could be considered critical or more advanced usage                                                                                                                                                                                                                  |

Reviewer 2: ```suggestion
| Additional library helper functions | <Badge text="0" variant="danger" /> | Helper and metadata accessor functions exported from the public surface of the compiler or a TypeSpec library are generally considered stable unless otherwise specified or only exported through an experimental subpath.                                                                                                                                                                                                               |
```

> | TypeSpec Syntax Tree              | <Badge text="2" variant="tip" />    | The TypeSpec AST is internal and may change at any time.                                                                                                                                                                                                                          |

Reviewer 2: ```suggestion
| TypeSpec AST (`Node`)             | <Badge text="2" variant="tip" />    | The TypeSpec abstract syntax tree is considered for internal use only and may change at any time. **Note**: this is not the same thing as the _type graph_ (`Type` and its variants), which is a Level 0, stable feature; this only applies to `Node` and its variants. Libraries depend on the abstract syntax tree _at their own risk_!                                                                                                                                                                                                                           |
```

> | Checker                          | <Badge text="2" variant="tip" />    | Checker is an internal of the typespec program                                                                                                                                                                                                                                                    |

Reviewer 2: ```suggestion
| Checker                          | <Badge text="2" variant="tip" />    | Direct use of the Program's type checker instance is considered for internal use only and its API may change at any time.                                                                                                                                                                                                                                                   |
```

> | Symbols                          | <Badge text="2" variant="tip" />    | Internals                                                                                                                                                                                                                                                                                         |

Reviewer 2: ```suggestion
| Symbols                          | <Badge text="2" variant="tip" />    | TypeSpec node symbols (not JavaScript symbols or state keys, rather the internal symbols used to relate TypeSpec types in the parser and type checker) are for internal use only.                                                                                                                                                                                                                                                                                    |
```

> | Formatter                        | <Badge text="1" variant="note" />   | Formatting shouldn't affect the meaning of the language. Changes to the formatting should be taken with consideration as it could create a lot of diff but it shouldn't also be considered a breaking change                                                                                     |

Reviewer 2: ```suggestion
| Formatter                        | <Badge text="1" variant="note" />   | Formatting shouldn't affect the meaning of the language. Changes to the formatting should be taken with consideration, as changes in formatting may introduce significant and unexpected differences in formatted output, but changes to the formatter output are not considered breaking changes for the purposes of issuing a major version change.                                                                                     |
```

> | Bug                              | [See bug](#bugs)                    | I think if something is found to be a bug from the documented behavior and fixing it is in someway breaking it shouldn't be considered as such unless it is such a widely used item where it has to become a "feature". Those should however still be documented as breaking in the release notes |

Reviewer 2: ```suggestion
| Bug fixes                             | [See bug](#bugs)                    | Bug fixes will not be considered breaking changes for the purposes of TypeSpec versioning if the existing behavior is _clearly_ broken or unintentional (e.g. if it violates the documented and expected behavior, or if there is no reasonable interpretation of the buggy behavior), even if the change is technically a breaking change. Bug fixes that are technically breaking will be documented as breaking changes in the release notes, but will not introduce a major version revision. However, if a bug's behavior has become so widely used as to be considered a "feature," we will consider the impact of the bugfix and make a reasonable determination to the best of our abilities and depending on the non-breaking mitigation strategies we have at our disposal. We will not break behaviors in stable releases that are known to be widely depended upon, even if we believe they are the result of a bug. |
```

> ### Supported node version
> 
> TypeSpec intends to support all [current node.js releases](https://nodejs.org/en/about/previous-releases). We may drop support for a maintenance or unsupported node version within a major version.

Reviewer 2: ```suggestion
### Supported Node.js Versions

TypeSpec will support all [current, active, and maintenance LTS Node.js releases](https://nodejs.org/en/about/previous-releases). We aim to support newly released current LTS versions as soon as possible, and we will provide a six month "grace period" after an LTS release leaves maintenance support before we cease testing it. After we drop support for a maintenance LTS version of Node.js, we may begin using syntax and runtime features that outdated Node.js versions may not support in new minor versions of published packages.

TypeSpec does not guarantee support for any alternative JavaScript engines or runtimes.
```

- TypeScript API types in TypeScript might change in ways that introduce TypeScript type checking errors (e.g. adding a new variant to a union). TypeScript type checking errors in library builds are not considered breaking changes, and library authors working in TypeScript should take care to program defensively as appropriate. More information about TypeScript API guarantees is provided in the section _Categories of Breaking Changes_ below.

### Bugs

A bug fix that introduces a technically breaking change will not be considered a breaking change for the purposes of TypeSpec versioning if it is clear that the existing behavior was broken or unintended in the first place.

## Categories of Breaking Changes

- <Badge text="Tier 0" variant="success" />: These are stable core APIs and language features that
  define fundamental semantics and will not be broken without a major version update. We recommend
  that library authors depend on these APIs.
- <Badge text="Tier 1" variant="caution" />: These are APIs that have very limited use cases and may
  be broken without a major version update. We recommend that library authors avoid depending on
  these APIs.

> - <Badge text="Tier 3" variant="danger" />: These are internal APIs intended only for use by the

Reviewer 3: Probably we want this to be tier 2 and the tier below to be tier 3.  (Later in this paragraph, this tier is described as tier 2)

>   TypeSpec compiler and core libraries. Tier 2 APIs may be changed without a a major version update

Reviewer 3: ```suggestion
  TypeSpec compiler and core libraries. Tier 2 APIs may be changed without a a major version update
```
Need to make sure this matches the Tier label

  or even documentation of the change. Consuming these APIs is _at your own risk_!

> - <Badge text="Tier 4" variant="note" />: These are APIs and language features that don't describe

Reviewer 3: I wonder if we should say something specific about emitter output for OpenAPI3.  For example, a non-guarantee that output will be textually identical but will be semantically equivalent.

Author: yeah thats a good point, i'll add an entry in the table

| Stdout                              | <Badge text="3" variant="danger" />  | TypeSpec does not provide any contract with the output emitted to `stdout`/`stderr`. The CLI might change the output at any time. Stderr is used to output progress tracking and should NOT be counted as a failure.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |

> | `@internal` exports                 | <Badge text="3" variant="danger" />  | Any function, class or JS component that is exported but doesn't have a TypeScript type due to be marked `@internal` is considered for internal use only and its API may change at any time.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |

Reviewer 3: ```suggestion
| `@internal` exports                 | <Badge text="3" variant="danger" />  | Any function, class or JS component that is exported but doesn't have a TypeScript type due to being marked `@internal` is considered for internal use only and its API may change at any time.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
```

| Formatter                           | <Badge text="4" variant="note" />    | Formatting shouldn't affect the meaning of the language. Changes to the formatting should be taken with consideration, as changes in formatting may introduce significant and unexpected differences in formatted output, but changes to the formatter output are not considered breaking changes for the purposes of issuing a major version change.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |

> | Bug fixes                           | [See bug](#bugs)                     | Bug fixes will not be considered breaking changes for the purposes of TypeSpec versioning if the existing behavior is _clearly_ broken or unintentional (e.g. if it violates the documented and expected behavior, or if there is no reasonable interpretation of the buggy behavior), even if the change is technically a breaking change. Bug fixes that are technically breaking will be documented as breaking changes in the release notes, but will not introduce a major version revision. However, if a bug's behavior has become so widely used as to be considered a "feature," we will consider the impact of the bugfix and make a reasonable determination to the best of our abilities and depending on the non-breaking mitigation strategies we have at our disposal. We will not break behaviors in stable releases that are known to be widely depended upon, even if we believe they are the result of a bug. |

Reviewer 3: I wonder if we should have a specific discussion channel for breaking change bugs

Author: yeah, I think we can see depending on how much we see it being a breaking issue and go through some more smooth rollout
